source/vir/src/modes.rs

```diff
@@ -32,7 +32,25 @@ pub fn mode_join(m1: Mode, m2: Mode) -> Mode {
     }
 }
 
-// placeholder for the erasure expected mode, which is computed after first mode checking pass
+// Placeholder for the erasure expected mode, which is computed after first mode checking pass.
+// (See in erase_expr(..., expect: Mode, ...) in erase.rs.)
+// We compute the erasure expected mode here so that the AIR/SMT generation can use it.
+// For example, if there are arithmetic operations that are exec (and therefore not erased),
+// then the AIR/SMT code will insert overflow checks for those arithmetic operations.
+// Example:
+//   fn test(#[exec] e: u64, #[spec] s: u64) {
+//     if e + 1 < s { ... }
+//   }
+// Here, e + 1 < y is erased because the "<" comparison with s is spec, not exec.
+// Therefore, the e + 1 is a spec-mode arithmetic operation, not an exec-mode arithmetic operation.
+// However, the mode checker doesn't know this when it first descends into e + 1:
+//   - the "if" expression can take a spec or proof or exec condition, so this doesn't constrain
+//     the mode of e + 1 < s a priori
+//   - the "<" expression can take spec or proof or exec subexpressions, so this doesn't constrain
+//     the mode of e + 1 a priori
+//   - we check the left side e + 1 before we check the right side s,
+//     so when we first reach e + 1, we don't yet know that "<"'s right side is spec
+// Therefore, we use Cell to delay the decision about the erasure expected mode of e + 1.
```

Comment:
so is it right to say that the "erasure expected mode" is really the actual "final decision" about the mode? 

And the mode returned by `check_expr` is more like a "pre-mode" or an approximation of the mode in some circumstances?

Reply:
The mode returned by `check_expr` is the tightest possible mode that the expression can have (if both exec and spec are possible, return exec).  Outer expressions might relax this mode (relax exec to proof, relax exec to spec, or relax proof to spec) via the "subsumption rule" described in https://github.com/Chris-Hawblitzel/rust/wiki/Three-kinds-of-code-...-specification,-proof,-and-executable .  This is like subtyping, where a type checker computes the most precise subtype for an expression, but might later relax this type to a supertype like `Object`.

Because operations like + and < can operate in different modes, there's more than one possible "final decision" that we could choose.  The erasure chooses a decision biased towards spec whenever possible in order to erase as much code as possible.  So you could call the erasure decision the "final decision" about the mode.  The checker doesn't care, but the erasure does.